dev/ToastNotifications/NotificationProperties.cpp

```diff
@@ -0,0 +1,139 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include "pch.h"
+
+#include "ToastNotification.h"
+#include "NotificationProperties.h"
+#include "ToastNotificationUtility.h"
+#include "../PushNotifications/PushNotificationUtility.h"
+
+namespace winrt
+{
+    using namespace winrt::Microsoft::Windows::ToastNotifications;
+}
+
+namespace Helpers
+{
+    using namespace winrt::Microsoft::Windows::PushNotifications::Helpers;
+}
+
+NotificationProperties::NotificationProperties(winrt::ToastNotification const& toastNotification)
+{
+    // Extract payload and convert it from XML to a byte array
+    auto payload = toastNotification.Payload();
+    auto payloadAsHstring = payload.GetXml();
+
+    std::wstring payloadAsStdWstring{ payloadAsHstring };
+
+    auto payloadAsSimpleString = Helpers::WideStringToUtf8String(payloadAsStdWstring);
+
+    m_payload = wil::unique_cotaskmem_array_ptr<byte>(static_cast<byte*>(CoTaskMemAlloc(payloadAsSimpleString.size())), payloadAsSimpleString.size());
+    CopyMemory(m_payload.data(), payloadAsSimpleString.c_str(), payloadAsSimpleString.size());
+
+    m_notificationId = toastNotification.ToastId();
+
+    m_tag = toastNotification.Tag();
+    m_group = toastNotification.Group();
+
+    auto expiry = toastNotification.ExpirationTime();
+    m_expiry = winrt::clock::to_file_time(expiry);
+
+    auto arrivalTime = winrt::clock::now();
+    m_arrivalTime = winrt::clock::to_file_time(arrivalTime);
+
+    m_expiresOnReboot = toastNotification.ExpiresOnReboot();
+
+    // TODO: Uncomment once ToastProgressData is implemented
+    // m_progressData = toastNotification.ProgressData();
+}
+
+STDMETHODIMP_(HRESULT __stdcall) NotificationProperties::get_NotificationId(_Out_ unsigned int* notificationId) noexcept
+{
+    auto lock = m_lock.lock_shared();
+
+    *notificationId = m_notificationId;
+    return S_OK;
+}
+
+STDMETHODIMP_(HRESULT __stdcall) NotificationProperties::get_Payload(_Out_ unsigned int* payloadSize, _Out_ byte** payload) noexcept
+{
+    auto lock = m_lock.lock_shared();
+
+    size_t tempPayloadSize = m_payload.size();
+
+    auto tempPayload = wil::unique_cotaskmem_array_ptr<byte>(static_cast<byte*>(CoTaskMemAlloc(tempPayloadSize)), tempPayloadSize);
+    CopyMemory(tempPayload.data(), m_payload.get(), tempPayloadSize);
+
+    *payloadSize = static_cast<unsigned int>(tempPayloadSize);
+    *payload = tempPayload.release();
+    return S_OK;
+}
+
+STDMETHODIMP_(HRESULT __stdcall) NotificationProperties::get_PayloadSize(_Out_ unsigned int* payloadSize) noexcept
+{
+    auto lock = m_lock.lock_shared();
+
+    *payloadSize = static_cast<unsigned int>(m_payload.size());
+    return S_OK;
+}
+
+STDMETHODIMP_(HRESULT __stdcall) NotificationProperties::get_Tag(_Out_ HSTRING* tag) noexcept
+{
+    auto lock = m_lock.lock_shared();
+
+    *tag = safe_make_unique_hstring(m_tag.c_str()).release();
```

Comment:
Let's do noexcept try catch

Reply:
in general all the APIs right?

Reply:
Only the ones that can throw, anything else can remain as noexcept.